mne/annotations.py

```diff
@@ -331,3 +331,140 @@ def _read_annotations(fid, tree):
+    regexp : str | None
+        Regular expression used to filter the annotations whose
+        descriptions is a match.
+    event_id_func : None | str | callable
+        What to do for events not found in ``event_id``. Must take one ``str``
+        argument and return an ``int``. If string, must be 'strip-to-integer',
+        in which case it defaults to stripping event codes such as "D128" or
+        "S  1" of their non-integer parts and returns the integer.
+        If the event is not in the ``event_id`` and calling ``event_id_func``
+        on it results in a ``TypeError`` (e.g. if ``event_id_func`` is
+        ``None``) or a ``ValueError``, the event is dropped.
+    convention : str
+        Can be "eeglab" or "default". XXX
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see
+        :func:`mne.verbose` and :ref:`Logging documentation <tut_logging>`
+        for more). Defaults to self.verbose.
+
+    Returns
+    -------
+    events : ndarray, shape (n_events, 3)
+        The events.
+    event_id : dict
+        The event_id variable that can be passed to Epochs.
+    """
+    import re
+
+    assert convention in ['default', 'eeglab']
+
+    if annotations is None:
+        annotations = raw.annotations
+
+    if convention == "default":
+        inds = raw.time_as_index(
+            annotations.onset, use_rounding=True) + raw.first_samp
+
+        unique_keys = np.unique(annotations.description)
+
+        logger.info('Annotations descriptions: %s' % unique_keys)
+
+        if regexp is not None:
+            pat = re.compile(regexp)
+            unique_keys = [kk for kk in unique_keys if pat.match(kk)]
+
+        if event_id is None:
+            event_id_ = dict(
+                zip(unique_keys, np.arange(1, 1 + len(unique_keys))))
+        else:
+            event_id_ = {k: v for k, v in event_id.items()
+                         if k in unique_keys}
+
+        event_sel = [ii for ii, kk in
+                     enumerate(annotations.description)
+                     if kk in event_id_]
+        if len(event_sel) == 0:
+            raise ValueError('Could not find any of the events you specified.')
+        values = [event_id_[kk] for kk in
+                  annotations.description[event_sel]]
+        previous_value = np.zeros(len(event_sel))
+        inds = inds[event_sel]
+        events = np.c_[inds, previous_value, values].astype(int)
+        return events, event_id_
+
+    elif convention == "eeglab":
+        types = annotations.description
+        latencies = annotations.onset
+
+        if "boundary" in types and "boundary" not in event_id:
+            warn("The data contains 'boundary' events, indicating data "
+                 "discontinuities. Be cautious of filtering and epoching "
+                 "around these events.")
+
+        if len(types) < 1:  # if there are 0 events, we can exit here
```

Comment:
This check should be before branching into conventions, I think. As I understand this: one scenerio is empty annotations attribute (`mne.annotations is None`) another is that the annotations are there but are empty (I guess this can only happend if one created empty annotations or added and then removed all annotations.

Reply:
This has to do with https://github.com/mne-tools/mne-python/pull/5416 But I'm not sure which is the best strategy. 

It is true that `xx.annotations is None` comes really natural. But then imagine the following scenario:

```py
@testing.requires_testing_data
def test_xx():

    #################################
    # Adapt the reference raw object
    #################################

    raw = read_raw_fif(fif_fname, verbose=0)
    raw_data_offset_in_sec = raw.first_samp / raw.info['sfreq']
    data_start_from_origin_in_sec = 50
    data_duration_in_sec = 10
    crop_tmin = data_start_from_origin_in_sec - raw_data_offset_in_sec
    crop_tmax = crop_tmin + data_duration_in_sec

    reference_mne_data = raw.copy().crop(tmin=crop_tmin, tmax=crop_tmax)

    # XXX There are more tests to be done to ensure that the reference object
    # is proper
    assert reference_mne_data.first_samp > raw.first_samp
    # Ensure first_samp
    expected_first_samp = data_start_from_origin_in_sec * raw.info['sfreq']
    assert reference_mne_data.first_samp == ceil(expected_first_samp) # XXX This bugs me

    assert reference_mne_data.annotations is None
    # Add some annotations
    orig_time = reference_mne_data.info['meas_date']
    reference_annotations = Annotations(onset=[51.2],
                                        duration=[1.0],
                                        description=['a'],
                                        orig_time=orig_time)
    reference_mne_data.set_annotations(reference_annotations)

    #################################
    # Split the data into 3 objects
    #################################

    split_duration_in_sec = data_duration_in_sec / 3
    part_a = reference_mne_data.copy().crop(tmin=0, tmax=split_duration_in_sec)
    part_b = reference_mne_data.copy().crop(tmin=split_duration_in_sec,
                                            tmax=2 * split_duration_in_sec)
    part_c = reference_mne_data.copy().crop(tmin=2 * split_duration_in_sec,
                                            tmax=None)

    # This seems a natural way to be called
    #
    assert part_a.annotations is not None
    assert part_b.annotations is None
    assert part_c.annotations is None


    # But I might also like to do that
    #
    n_annotations = [p.annotations.length() for p in (part_a, part_b, part_c)]
    assert n_annotations == [1, 0, 0]
```

you either have one or the other. But you cannot have both, because the `is` operator cannot be overwritten.

Reply:
And the problem is that once you assign to `None` you are no longer an `Annotation`. 

The concept of `None` is not an empty `Annotation` but a nothing object.

and I don't seem to be able to figure out which is the nicest of the two approaches.

Reply:
@massich I understand your concerns, in that case empty annotations would be ok. I can still easily check `p.annotations.length() > 0` instead of checking if they are None.